linker.go

```diff
@@ -57,10 +57,24 @@ func splitSymbols(insns asm.Instructions) (map[string]asm.Instructions, error) {
 		return nil, errors.New("insns is empty")
 	}
 
-	if insns[0].Symbol() == "" {
+	firstSymbol := insns[0].Symbol()
+	if firstSymbol == "" {
 		return nil, errors.New("insns must start with a Symbol")
 	}
 
+	onlyOneSymbol := true
+	for _, ins := range insns[1:] {
+		if ins.Symbol() != "" {
+			onlyOneSymbol = false
+			break
+		}
+	}
+	if onlyOneSymbol {
+		return map[string]asm.Instructions{
+			firstSymbol: insns,
```

Comment:
Currently splitSymbols always returns a copy of insns. The new code will sometimes return a copy, sometimes not, which is a great bug magnet IMO.

Have you tried instead to optimize the loop below? I can see two big inefficiencies:

* append will do extra copies
* map lookup and assigns are also expensive

Getting rid of the two should make this a lot better without changing the alloc behaviour. You can probably track last name + index and do an exact allocation + copy whenever name changes.

Reply:
That's a good point, I checked that we were not modifying the result of this function but this can change in the future.

I think the main point is that we split into a map and then iterate over the map to create another array (in the only place this function is called). Would you be open to removing the map completely and replacing it with an iterator so that we skip the intermediate allocation completely ?

Reply:
Basically `splitSymbols` returns a map, but we never really take advantage of the fact that it's a map (except in tests), we directly iterate on it and create something else from it. What I was suggesting is that `splitSymbols` returns a lazy iterator type that would allow iteration without creating a big map with all the symbols and instructions.

I don't think it matters that much, in the latest version of the commit I improved the algorithm a lot in both the fast and slow paths.